Foundation/NSXMLNode.swift

```diff
@@ -50,291 +51,832 @@ public enum NSXMLNodeKind : UInt {
+
+        case _kCFXMLTypeDocument:
+            return NSXMLDocument._objectNodeForNode(node)
+
+        case _kCFXMLTypeDTD:
+            return NSXMLDTD._objectNodeForNode(node)
+
+        default:
+            let _private = _CFXMLNodeGetPrivateData(node)
+            if _private != nil {
+                let unmanaged = Unmanaged<NSXMLNode>.fromOpaque(_private)
+                return unmanaged.takeUnretainedValue()
+            }
+
+            return NSXMLNode(ptr: node)
+        }
+    }
+
+    // libxml2 believes any node can have children, though NSXMLNode disagrees.
+    // Nevertheless, this belongs here so that NSXMLElement and NSXMLDocument can share
+    // the same implementation.
+    internal func _insertChild(child: NSXMLNode, atIndex index: Int) {
+        precondition(index >= 0)
+        precondition(index <= childCount)
+        precondition(child.parent == nil)
+
+        _childNodes.insert(child)
+
+        if index == 0 {
+            let first = _CFXMLNodeGetFirstChild(_xmlNode)
+            _CFXMLNodeAddPrevSibling(first, child._xmlNode)
+        } else {
+            let currChild = childAtIndex(index - 1)!._xmlNode
+            _CFXMLNodeAddNextSibling(currChild, child._xmlNode)
+        }
+    } //primitive
+
+    // see above
+    internal func _insertChildren(children: [NSXMLNode], atIndex index: Int) {
+        for (childIndex, node) in children.enumerate() {
+            _insertChild(node, atIndex: index + childIndex)
+        }
+    }
+
+    /*!
+     @method removeChildAtIndex:atIndex:
+     @abstract Removes a child at a particular index.
+     */
+    // See above!
+    internal func _removeChildAtIndex(index: Int) {
+        guard let child = childAtIndex(index) else {
+            fatalError("index out of bounds")
+        }
+
+        _childNodes.remove(child)
+        _CFXMLUnlinkNode(child._xmlNode)
+    } //primitive
+
+    // see above
+    internal func _setChildren(children: [NSXMLNode]?) {
+        _removeAllChildren()
+        guard let children = children else {
+            return
+        }
+
+        for child in children {
+            _addChild(child)
+        }
+    } //primitive
+
+    /*!
+     @method addChild:
+     @abstract Adds a child to the end of the existing children.
+     */
+    // see above
+    internal func _addChild(child: NSXMLNode) {
+        precondition(child.parent == nil)
+
+        _CFXMLNodeAddChild(_xmlNode, child._xmlNode)
+        _childNodes.insert(child)
+    }
+
+    /*!
+     @method replaceChildAtIndex:withNode:
+     @abstract Replaces a child at a particular index with another child.
+     */
+    // see above
+    internal func _replaceChildAtIndex(index: Int, withNode node: NSXMLNode) {
+        let child = childAtIndex(index)!
+        _childNodes.remove(child)
+        _CFXMLNodeReplaceNode(child._xmlNode, node._xmlNode)
+        _childNodes.insert(node)
+    }
 }
 
+extension NSXMLNode: CollectionType {
```

Comment:
So the thing is, I love this idea, the way it so neatly fit CollectionType with so little code. And I use it elsewhere in the implementation. I would have marked it as `internal`, but of course we can't use access control on extensions that declare protocol conformance, so I couldn't. Should I get rid of it and refactor the code that uses it, or should I put the "experimental" comment on it, since it's purely additive and doesn't change or replace any existing API?